cpu/numactl.py

```diff
@@ -238,14 +238,16 @@ def test_preferred_node(self):
                                                 self.bytes,
                                                 self.count)
                 self.dd_run(cmd)
+        else:
+            self.cancel("Device input missing, skipping the test")
 
     def test_cpunode_with_membind(self):
         '''
         Test CPU and memory bind
         '''
         if not self.iface and not self.disk:
             self.cancel("Network inferface or disk/device input missing")
-        if self.check_numa_nodes():
+        if self.device and self.check_numa_nodes():
```

Comment:
@AYUSHJAIN951 why can't we cancel this in the setup itself , which will be a just else case and why do you want to do it in the test?
Please cancel it in the setup itself.

Reply:
There are 7 testcases totally and first 3 does not need self.device input and can run without that as well,
So Cancelling test cases which require device parameter only make more sense rather than Cancelling in setup itself, and not letting first 3 tests also to run.

Reply:
@AYUSHJAIN951 check line 124-126 in setup. it is already handled there.

```
 if self.device:
            if not os.path.isdir('/sys/bus/pci/devices/%s' % self.device):
                self.cancel("%s not present in device path" % self.device)
```

Reply:
I feel when this is a complete test and few subtest uses pci_device and few doesn't then also we have cancel the test in setup itself because our main aim is run the complete test and not only few sub-tests, and this test mean for it.

@PraveenPenguin @narasimhan-v ^^

Reply:
When using [numa_fc.yaml](https://github.com/avocado-framework-tests/avocado-misc-tests/blob/bf81179a0a7c218363d2dcdc0fb4b8211541e276/cpu/numactl.py.data/numa_fc.yaml).
There is no Interface parameter and as you correctly pointed out, that there is a Cancel in the setup itself, 
But it is only when interface is provided by the user not the other way around.

So if no interface parameter than again we cant cancel in the setup itself for all tests,
as we have a possibility to continue test for few subtests and only fail when needed.

Reply:
I agree we should cancel in setup. We can use self.test to get test name and cancel for those tests in setup

Reply:
@AYUSHJAIN951 Please change accordingly.

Reply:
@Naresh-ibm Can you Please review again as now its being handled in setup stage itself.

Reply:
Ok, Thanks for changes @AYUSHJAIN951. can you please attach the changed logs?

Reply:
[cpu_numactl.txt](https://github.com/avocado-framework-tests/avocado-misc-tests/files/11310044/cpu_numactl.txt)
Logs!